Bluetooth: Userchan: Parse multiple hci packets

vChavezB opened the pull request:
After conducting tests with a a virtual Bluetooth controller over TCP it was noticed that some HCI packets may arrive on the same buffer if sent over a short period of time.
This update ensures the hci packets are parsed correctly in the case multiple packets are in the same recieving buffer according to how HCI packets are decoded in the Bluetooth Spec v5.2 Part E.

This should fix https://github.com/zephyrproject-rtos/zephyr/issues/61762

Open for suggestions and discussion on how to address this the best way.

Commit:
Bluetooth: controller: Added HCI SCO header definition

Added HCI SCO header definition and included comments
that refer to the Bluetooth core spec sections for
the other HCI header types.

Signed-off-by: Victor Chavez <chavez-bermudez@fh-aachen.de>

Thalley commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +93,53 @@ static struct net_buf *get_rx(const uint8_t *buf)
+	/* Bluetooth spec 5.2 Vol 4, Part E - 5.4.2 ACL Data Packets */
+	case H4_ACL:
+		payload_size = packet_buff[4] << 8 | packet_buff[3];
+		header_size = 4;
```
```suggestion
		header_size = BT_HCI_ACL_HDR_SIZE;
```
or
```suggestion
		header_size = sizeof(struct bt_hci_acl_hdr);
```

vChavezB replied:
Thanks for bringing this up. Just a remark...for the HCI H4 SCO, I did not find a definition, should I add it to https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/bluetooth/hci_types.h ? 

e.g., 

```c
#define BT_HCI_SCO_HDR_SIZE 3
```

Thalley replied:
Yes, please add the `BT_HCI_SCO_HDR_SIZE` along side the actual header - Something similar to 
```c
struct bt_hci_acl_hdr {
	uint16_t handle;
	uint16_t len;
} __packed;
```

Thalley commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +93,53 @@ static struct net_buf *get_rx(const uint8_t *buf)
+	switch (packet_type) {
+	/* Bluetooth spec 5.2 Vol 4, Part E - 5.4.1 HCI Command Packet */
+	case H4_CMD:
+		payload_size = packet_buff[3];
+		header_size = 3;
+		break;
```
There is some information missing from when you are doing `[3]` or `[2]`. I assume that is has something to do with the header sizes? 

Suggest to use defined values, rather than magic numbers, to retrieve the payload sizes

Thalley commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -134,27 +181,45 @@ static void rx_thread(void *p1, void *p2, void *p3)
+			buf_tailroom = net_buf_tailroom(buf);
+			buf_add_len = decoded_len-1;
+			if (buf_tailroom < buf_add_len) {
+				LOG_ERR("Not enough space in buffer %zu/%zu",
+						buf_add_len, buf_tailroom);
```
```suggestion
				LOG_ERR("Not enough space in buffer %zu/%zu",
					buf_add_len, buf_tailroom);
```

Thalley commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +100,56 @@ static struct net_buf *get_rx(const uint8_t *buf)
+	uint8_t header_size = 0;
+	const uint8_t packet_type_len = 1;
+
+	switch (packet_type) {
+	/* Bluetooth spec 5.2 Vol 4, Part E - 5.4.1 HCI Command Packet */
+	case H4_CMD:
+		/* Parameter Total Length */
+		payload_size = packet_buff[H4_CMD_PAYLOAD_SIZE_OFF];
+		header_size = BT_HCI_CMD_HDR_SIZE;
+		break;
+	/* Bluetooth spec 5.2 Vol 4, Part E - 5.4.2 ACL Data Packets */
+	case H4_ACL:
+		/* Data Total Length */
+		payload_size = sys_get_le16(&packet_buff[H4_ACL_PAYLOAD_SIZE_OFF]);
+		header_size = BT_HCI_ACL_HDR_SIZE;
+		break;
+	/* Bluetooth spec 5.2 Vol 4, Part E - 5.4.3 HCI Synchronous Data Packets */
+	case H4_SCO:
+		/* Data_Total_Length */
+		payload_size = packet_buff[H4_SCO_PAYLOAD_SIZE_OFF];
+		header_size = 3;
+	/* Bluetooth spec 5.2 Vol 4, Part E - 5.4.4 HCI Event Packet */
+	case H4_EVT:
+		/* Data_Total_Length */
+		payload_size = packet_buff[H4_EVT_PAYLOAD_SIZE_OFF];
+		header_size = BT_HCI_EVT_HDR_SIZE;
+		break;
+	/* Bluetooth spec 5.2 Vol 4, Part E - 5.4.5 HCI ISO Data Packets */
+	case H4_ISO:
+		/* ISO_Data_Load_Length parameter */
+		payload_size =  sys_get_le16(&packet_buff[H4_ISO_PAYLOAD_SIZE_OFF]) & 0x3FFF;
+		header_size = BT_HCI_ISO_HDR_SIZE;
```
I wonder if it would make more sense to actually extract the header in each case, and use the `len` field in each header to extract the `payload_size`. That might be better than using these offsets and so on. 

So it would be something like 

```c
	case H4_ACL:
		/* Data Total Length */
		const struct bt_hci_acl_hdr *hdr = &packet_buff[sizeof(packet_type)];

		payload_size = sys_get_le16(hdr->len);
		header_size = sizeof(hdr);
		break;
```

Though it may require some endianess handling or something for this to properly work.

vChavezB replied:
Yes 👍 , I was checking the source code of how the hci structs are handled and its similar to what you mention

See for example

https://github.com/zephyrproject-rtos/zephyr/blob/2d65acca3a76b968657c2e796512558e93f2ed23/drivers/bluetooth/hci/h4.c#L127C3-L127C3.

Thalley commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +100,56 @@ static struct net_buf *get_rx(const uint8_t *buf)
+{
+	uint16_t payload_size = 0;
+	uint8_t header_size = 0;
+	const uint8_t packet_type_len = 1;
```
```suggestion
	const uint8_t packet_type_len = sizeof(packet_type);
```

jhedberg commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +95,61 @@ static struct net_buf *get_rx(const uint8_t *buf)
+		header_size = BT_HCI_SCO_HDR_SIZE;
+		break;
+	case H4_EVT:
+		const struct bt_hci_evt_hdr *hdr = &packet_buff;
```
This doesn't seem correct. You have the packet type as the first byte and the packet-specific header comes after that. In all other cases you read `&packet_buff[1]` but here you read `&packet_buff[0]`. Actually, I don't see any reason why you need two parameters to this function. Just pass the original buffer and inside the function you can then do `uint8_t type = buf[0];`.

vChavezB replied:
Yes, sorry mistake on my part while changing to the suggestions of the other reviewer.

jhedberg commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +95,61 @@ static struct net_buf *get_rx(const uint8_t *buf)
+		const struct bt_hci_acl_hdr *hdr = &packet_buff[sizeof(packet_type)];
+
+		/* Data Total Length */
+		payload_size = sys_get_le16(&hdr->len);
```
This generates unnecessarily complicated code. Just do `size = sys_le16_to_cpu(hdr->len);`. The structs are packed, so accessing the struct members generates automatically alignment-safe code (if needed by the CPU architecture).

vChavezB replied:
Does the packed attribute deal with endianness ? Thats the reason I thought parts like these ones are needed in the current bluetooths stack 

https://github.com/zephyrproject-rtos/zephyr/blob/2d65acca3a76b968657c2e796512558e93f2ed23/drivers/bluetooth/hci/h4.c#L127C3-L127C3

https://github.com/zephyrproject-rtos/zephyr/blob/2d65acca3a76b968657c2e796512558e93f2ed23/drivers/bluetooth/hci/h4.c#L140C2-L140C2

vChavezB replied:
I changed the other comment reviews accordingly but this one is still is not clear to me.

Could you be more specific on your suggestion? 

>  Just do `size = sys_le16_to_cpu(hdr->len)`

I understood either:

1. Should I change the name of the variable `payload_len` to `size` ?
2. Should I obtain the size of the payload via `size = sys_le16_to_cpu(hdr->len);` before the switch case? Its a trivial change and only the edge case for the struct `bt_hci_cmd_hdr` has to be taken into account (`param_len`).


Sorry for the misunderstanding on my part, looking forward for your comments.

vChavezB replied:
Ok, I think I overthought the comment. Changed to using `sys_le16_to_cpu` instead of `sys_get_le16` and the lenght member of the respective hci header struct instead of the address.

jhedberg commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +95,61 @@ static struct net_buf *get_rx(const uint8_t *buf)
+{
+	uint16_t payload_size = 0;
+	uint8_t header_size = 0;
+	const uint8_t packet_type_len = sizeof(packet_type);
```
This seems like an unnecessary variable (doesn't add much to readability, and the way you've named it it's even longer than`sizeof(packet_type)`.

jhedberg commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +95,61 @@ static struct net_buf *get_rx(const uint8_t *buf)
+ */
+static uint16_t decode_hci_packet_length(uint8_t packet_type, const uint8_t *packet_buff)
+{
+	uint16_t payload_size = 0;
+	uint8_t header_size = 0;
```
Call these `len` instead of `size`.

jhedberg commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +95,61 @@ static struct net_buf *get_rx(const uint8_t *buf)
+ * @param packet_type	Type of packet according to Bluetooth spec v5.2 Vol 4, Part A Table 2.1
+ * @param packet_buff	Pointer to packet buffer
+ */
+static uint16_t decode_hci_packet_length(uint8_t packet_type, const uint8_t *packet_buff)
```
I'd just call this function `packet_len()`

jhedberg commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -37,6 +37,14 @@ extern "C" {
 #define BT_HCI_PAWR_SUBEVENT_MAX 128
 
+/* Bluetooth spec 5.2 Vol 4, Part E - 5.4.3 HCI Synchronous Data Packets */
+struct bt_hci_sco_hdr {
+	uint16_t  conn_handle;
+	uint8_t  len;
+} __packed;
```
Please follow the same struct alignment style as elsewhere in the headers, so that the struct member names are aligned. That means two spaces after `uint8_t` but only one after `uint16_t`.

vChavezB replied:
changed

Thalley commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +95,60 @@ static struct net_buf *get_rx(const uint8_t *buf)
+{
+	uint16_t payload_len = 0;
+	uint8_t header_len = 0;
+	const uint8_t packet_type = packet_buff[0];
+
+	switch (packet_type) {
+	case H4_CMD:
+		const struct bt_hci_cmd_hdr *hdr = &packet_buff[sizeof(packet_type)];
```
```suggestion
	const uint8_t packet_type = packet_buff[0];
	const uint8_t *packet_hdr = &packet_buff[sizeof(packet_type)];

	switch (packet_type) {
	case H4_CMD:
		const struct bt_hci_cmd_hdr *hdr = packet_hdr;
```

Since they all share the same offset, consider making doing it a single time.

vChavezB commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -37,6 +37,14 @@ extern "C" {
 #define BT_HCI_SYNC_HANDLE_INVALID 0xffff
 #define BT_HCI_PAWR_SUBEVENT_MAX 128
 
+/* Bluetooth spec 5.2 Vol 4, Part E - 5.4.3 HCI Synchronous Data Packets */
```
Changed. Checked the numbering of the sections in the sepc and they are the same so I just changed the text to v5.4

Thalley commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +95,67 @@ static struct net_buf *get_rx(const uint8_t *buf)
+	case H4_CMD: {
+		const struct bt_hci_cmd_hdr *hdr = packet_hdr;
+
+		/* Parameter Total Length */
+		payload_len = hdr->param_len;
+		header_len = BT_HCI_CMD_HDR_SIZE;
+		break;
```
Final comment: Just a thought, not a request:
We could also just do
```suggestion
		return sizeof(packet_type) + BT_HCI_CMD_HDR_SIZE + hdr->param_len;
```

and similar for all cases, to omit the 2 local variables (`payload_len` and `header_len`). This could reduce the size of this function by 14 lines 

Up to you if you want to change this, or leave it as is

jhedberg commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +95,67 @@ static struct net_buf *get_rx(const uint8_t *buf)
+ * @param packet_buff	Pointer to a HCI packet buffer
+ * @return Length of the HCI packet in bytes, zero if no valid packet found.
+ */
+static uint16_t packet_len(const uint8_t *packet_buff)
```
Sorry if this is getting a bit overly pedantic, but I'd prefer if we stick to existing variable naming conventions and also avoid unnecessarily long names. This helps keep the code more readable (for someone who has read a lot of other Bluetooth subsystem code). and also avoids excessively long lines.

Just call the input parameter `const uint8_t *buf`

vChavezB replied:
Oh no problem, I am aware that style and naming conventions are highly dependent on organisations policies. As I am still not that familiar with the codebase of Zephyr I know this could come up.

jhedberg commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +95,67 @@ static struct net_buf *get_rx(const uint8_t *buf)
+{
+	uint16_t payload_len = 0;
+	uint8_t header_len = 0;
+	const uint8_t packet_type = packet_buff[0];
+	const uint8_t *packet_hdr = &packet_buff[sizeof(packet_type)];
```
```c
const uint8_t type = buf[0];
const uint8_t hdr = &buf[sizeof(type)];
```

vChavezB replied:
changed

jhedberg commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +95,67 @@ static struct net_buf *get_rx(const uint8_t *buf)
+
+	switch (packet_type) {
+	case H4_CMD: {
+		const struct bt_hci_cmd_hdr *hdr = (const struct bt_hci_cmd_hdr *)packet_hdr;
```
```c
const struct bt_hci_cmd_hdr *cmd = (const struct bt_hci_cmd_hdr *)hdr;
```

jhedberg commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +95,67 @@ static struct net_buf *get_rx(const uint8_t *buf)
+		break;
+	}
+	case H4_ACL: {
+		const struct bt_hci_acl_hdr *hdr = (const struct bt_hci_acl_hdr *)packet_hdr;
```
```c
const struct bt_hci_acl_hdr *acl = (const struct bt_hci_acl_hdr *)hdr;
```

vChavezB replied:
changed

jhedberg commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +95,67 @@ static struct net_buf *get_rx(const uint8_t *buf)
+		break;
+	}
+	case H4_SCO: {
+		const struct bt_hci_sco_hdr *hdr = (const struct bt_hci_sco_hdr *)packet_hdr;
```
```c
const struct bt_hci_sco_hdr *sco = (const struct bt_hci_sco_hdr *)hdr;
```

jhedberg commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +95,67 @@ static struct net_buf *get_rx(const uint8_t *buf)
+		break;
+	}
+	case H4_EVT: {
+		const struct bt_hci_evt_hdr *hdr = (const struct bt_hci_evt_hdr *)packet_hdr;
```
```c
const struct bt_hci_evt_hdr *evt = (const struct bt_hci_evt_hdr *)hdr;
```

vChavezB replied:
changed

jhedberg commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +95,67 @@ static struct net_buf *get_rx(const uint8_t *buf)
+		break;
+	}
+	case H4_ISO: {
+		const struct bt_hci_iso_hdr *hdr = (const struct bt_hci_iso_hdr *)packet_hdr;
```
```c
const struct bt_hci_iso_hdr *iso = (const struct bt_hci_iso_hdr *)hdr;
```

vChavezB replied:
changed

jhedberg commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +95,67 @@ static struct net_buf *get_rx(const uint8_t *buf)
+		header_len = BT_HCI_ISO_HDR_SIZE;
+	}
+	/* If no valid packet type found */
+	default:
+		return 0;
```
Might make sense to add a `LOG_WRN("Unknown packet type 0x%02x", type)` before the return.

vChavezB replied:
added

jhedberg reviewed (changes requested):
One more review round (thanks for your patience!). After that I think this is ready to be merged.

I think it'd also make sense to split the `hci_types.h` changes into its own commit (still in this same PR) before the `userchan.c` commit, since those changes are not strictly related to the userchan driver.

jhedberg commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -37,6 +37,14 @@ extern "C" {
 
+/* Bluetooth spec v5.4 Vol 4, Part E - 5.4.3 HCI Synchronous Data Packets */
+struct bt_hci_sco_hdr {
+	uint16_t conn_handle;
```
Ok, I just checked and the HCI spec uses `Connection_Handle` both for SCO and ISO, but just `Handle` for ACL. However, since the ISO struct already uses simply `handle` I'd follow that convention for SCO as well.

vChavezB replied:
changed

vChavezB commented:
Added last suggestions. Pending separation of commits...

Thalley commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +95,68 @@ static struct net_buf *get_rx(const uint8_t *buf)
+ * @param buff	Pointer to a HCI packet buffer
+ * @return Length of the HCI packet in bytes, zero if no valid packet found.
+ */
+static uint16_t packet_len(const uint8_t *buff)
```
```suggestion
static uint16_t packet_len(const uint8_t *buf)
```

vChavezB replied:
good catch, changed

Commit:
Bluetooth: Userchan: Parse multiple hci packets

After conducting tests with a a virtual Bluetooth controller
over TCP it was noticed that some HCI packets may arrive on the
same buffer if sent over a short period of time.
This update ensures the hci packets are parsed correctly in the case
multiple packets are in the same recieving buffer according to
the Bluetooth Spec v5.4 Part E.

Signed-off-by: Victor Chavez <chavez-bermudez@fh-aachen.de>

vChavezB commented:
I have separated the commits.

Thalley commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -93,6 +95,68 @@ static struct net_buf *get_rx(const uint8_t *buf)
+		const struct bt_hci_iso_hdr *iso = (const struct bt_hci_iso_hdr *)hdr;
+
+		/* ISO_Data_Load_Length parameter */
+		payload_len =  bt_iso_hdr_len(sys_le16_to_cpu(iso->len));
+		header_len = BT_HCI_ISO_HDR_SIZE;
+	}
```
```suggestion
		payload_len =  bt_iso_hdr_len(sys_le16_to_cpu(iso->len));
		header_len = BT_HCI_ISO_HDR_SIZE;
		break;
	}
```

vChavezB replied:
Thanks for noticing this, have been changing different parts that I must have deleted the break while refactoring.

Thalley replied:
Yeah, it existed earlier :D
